tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -332,7 +374,151 @@ TEST_F(OverlapTest, ReduceScatterBasedPipeliningHostIrImplementation) {
 
     hie.runWithInput(std::move(inputs));
   }
-  validate();
 }
 
+using RingBasedOverlapTestParams = std::tuple<
+    bool, // use_coalescence
+    CommunicatorBackend>;
+
+class RingBasedOverlapTest
+    : public OverlapTest,
+      public testing::WithParamInterface<RingBasedOverlapTestParams> {
+ protected:
+  bool use_coalescence_;
+  at::Tensor src_buffer_, dst_buffer_;
+  void SetUp() override {
+    use_coalescence_ = std::get<0>(GetParam());
+    params.backend_type = std::get<1>(GetParam());
+
+    OverlapTest::SetUp();
+
+    std::vector<int64_t> buffer_sizes = {
+        params.S, params.M / params.S, params.N};
+    src_buffer_ = at::empty(buffer_sizes, gpu_options_);
+    dst_buffer_ = at::empty(buffer_sizes, gpu_options_);
+  }
+
+  at::Tensor getExpectedResult() override {
+    auto tc_unsharded_expected = getUnshardedExpectedResult();
+    // the natural layout here differs from the collective based pipelining.
+    // Here, the output is sharded on the outermost axis whereas, in the
+    // collective based pipelining, it is sharded on axis(1). The two layouts
+    // coincide in the classical case where params.S = num_devices_
+    auto tc_unsharded_expected_reshaped = at::reshape(
+        tc_unsharded_expected,
+        {num_devices_, params.S / num_devices_, params.M / params.S, params.N});
+    auto tc_expected =
+        tc_unsharded_expected_reshaped.select(0, my_device_index_);
+    return tc_expected.reshape(
+        {params.S, params.M / (params.S * num_devices_), params.N});
+  }
+};
+
+TEST_P(
+    RingBasedOverlapTest,
+    ReduceScatterRingBasedPipeliningATenImplementation) {
+  std::vector<c10::cuda::CUDAStream> streams =
+      createStreams(params.number_of_streams, my_device_index_);
+
+  ASSERT_EQ(params.S % num_devices_, 0);
+  int64_t& number_of_steps_per_ring = num_devices_;
+  int64_t number_of_rings = params.S / num_devices_;
+
+  auto ta_reshaped = at::reshape(
+      ta_,
+      {number_of_steps_per_ring,
+       number_of_rings,
+       params.M / params.S,
+       params.K / num_devices_});
+  auto src_buffer_reshaped = at::reshape(
+      src_buffer_,
+      {number_of_steps_per_ring,
+       number_of_rings,
+       params.M / params.S,
+       params.N});
```

Comment:
This is as large as the whole output matrix? Probably OK for the first version. However, I believe there's a way to reuse chunks by limiting the number of streams and putting extra synchronization, e.g., https://docs.google.com/document/d/1BK_xEolbJDEky5trLW8jbuNHMX-v2QSY40O8782lYQ8/edit#heading=h.g5s0igtg8my1

Reply:
Yes, this is the note I wrote at the end of the PR description under the header "Reducing memory foot print"